notebooks/data_table_transformer/main.py

```diff
@@ -0,0 +1,125 @@
+    - `value_columns` long format columns to carry over into the wide formatted table (These may be muxed with the
+       `header_column` names, e.g. "cram" -> "file_type-cram")
+
+    While this notebook is focused on the AnVIL data model, it can work with any Terra data table with appropriate
+    choices for `index_column`, `header_column`, and `value_columns`.
+    """
+
+with herzog.Cell("markdown"):
+    """
+    Install the newest version of [terra-notebook-utils](https://github.com/DataBiosphere/terra-notebook-utils)
+    and [terra-pandas](https://github.com/xbrianh/terra-pandas)
+    """
+
+with herzog.Cell("python"):
+    #%pip install --upgrade --no-cache-dir terra-notebook-utils
+    #%pip install --upgrade --no-cache-dir terra-pandas
+    pass
+
+with herzog.Cell("markdown"):
+    """
+    Import Pandas interface functions from [terra-pandas](https://github.com/xbrianh/terra-pandas)
+    """
+
+with herzog.Cell("python"):
+    from terra_pandas import table_to_dataframe, dataframe_to_table, long_to_wide
+
+with herzog.Cell("markdown"):
+    """
+    Carry out the long to wide data table transformation, which involves several steps:
+      - Convert the Terra data table into a Pandas dataframe with `table_to_dataframe`
+      - Filter out rows that cannot be transformed into long format. In this example we are transforming the data table
+        into a wide format using sample id. Multisample files, such as jointly called VCFs, will not fit into the
+        transformed dataframe. There are also extra complexities of the AnVIL data model that need to be addressed,
+        as explained below.
+      - Convert the wide-formatted dataframe into a Terra data table.
+    """
+
+import json
+import pandas
+from uuid import uuid4
+from random import choice, randint
+from terra_notebook_utils import table
+
+NUMBER_OF_WIDE_ROWS = 20
+
+long_table = "test-long"
+wide_table = "test-wide"
+
+table.delete(long_table)
+table.delete(wide_table)
+
+lines = list()
+for _ in range(NUMBER_OF_WIDE_ROWS):
+    sample_id = f"{uuid4()}"
+    sample_obj = dict(foo=f"{uuid4()}", entityName=sample_id)
+    for fmt in ["cram", "crai", "vcf"]:
+        lines.append({'pfb:sample': sample_obj,
+                      'pfb:data_format': fmt,
+                      'pfb:object_id': f"{uuid4()}",
+                      'pfb:file_size': randint(1024, 1024 ** 3),
+                      'pfb:file_name': f"{uuid4()}"})
+dataframe_to_table("test-long", pandas.DataFrame(lines))
+
+with herzog.Cell("python"):
+    index_column = "pfb:sample"
+    header_column = "pfb:data_format"
+    value_columns = ["pfb:object_id", "pfb:file_size", "pfb:file_name"]
+
+    # Uncomment the following lines with your table names
+    # long_table = ""  # Include long formatted table to transform
+    # wide_table = ""  # Include the wide formatted table to create
+
+    # Convert the long table into a Pandas dataframe
+    df = table_to_dataframe(long_table)
```

Comment:
Taken by itself, `long_table` would be an undeclared variable here, and appears to work by taking advantage of the testing above:

```
long_table = "test-long"
wide_table = "test-wide"
```

However, this breaks reverse translation compatibility, as cells in a `.pynb` file do not include the testing above, and a round trip would lose this and appear to create a failing notebook.

Consider explicitly declaring `long_table = "test-long"` and `wide_table = "test-wide"` here with a comment to the user to change these?

If this is not a concern, feel free to ignore this comment.

Reply:
This notebook is not runnable without user input, and cannot be made so. For testing, it takes advantage of the ability to declare functionality outside of the context of cell definitions. There may be a better way to request user input, I’m open to suggestions! I think it’s preferable for the notebook to error if use input is not provided.

Tests and fixtures are lost in the forward translation, and are not available for reverse translation. Forward translation is the main feature of herzog. Reverse translation is a convenience to bring notebooks into the herzog format. We should not aim for perfect round trip fidelity.